RFC-0009-jdbc-join-push-down.md

```diff
@@ -0,0 +1,798 @@
+
+Now we have new TableScanNode with list of joinTables. The split manager works on JdbcSplit and provide objects for processing on connector level. We need to modify the logic to transfer the new object (Optional<List<ConnectorTableHandle>> joinPushdownTables) too, to the connector level. Currently, for the Join query, the split is transfer to buildSql() to create the select statement. If the split contains ‘joinTables’ details then we need to transfer this ‘joinTables’ details to the new method called ‘buildJoinSql()’ where we need to create and return join query instead of select query.
+
+On buildJoinSql(), we need to handle columns to be selected, the tables from which to query, the join condition and filter conditions if any.
+
+#### 1. Handling Select columns
+
+The select column needs to resolve from the existing input argument List<JdbcColumnHandle> columns
+
+#### 2. Handling From tables.
+
+The parameter 'List<ConnectorTableHandle> joinTables' that is received on buildJoinSql() contains the tables in an order to join.  We can use this list of tables as from tables.
+
+#### 3. Handling JoinType
+
+We need not consider the JoinType, but we need to write the join query in the form of where clause. 
+
+#### 4. Handling Join criteria
+
+The Join criteria is also available in Optional<JdbcExpression> additionalPredicate.
+
+#### 5. Resolving Assignment
+
+The select column name and join criteria may be available as an expression and so each column (expression) we need to resolve from exact table assignement. Against each  JoinTables we have assignments object and we need to extract the actual table name and column name from this assignment for each select and criteria column.
+
+#### 6. Handling Filter
+
+There is no change expected but we may need to handle the assignment and alias.
+
+## Session flag
+
+#### 1. Enable JdbcJoinPushdown at session level 
+
+We have a new session flag 'optimizer.inner-join-pushdown-enabled'. This flag should be configured in presto-main config.properties with default value as false.
+eg:
+```
+optimizer.inner-join-pushdown-enabled = true
+```
+It can be set from user session to override the above config.
+eg:
+```
+SET SESSION optimizer_inner_join_pushdown_enabled = true
+```
+If we do not set this flag or set it to false (SET SESSION optimizer_inner_join_pushdown_enabled = false’) then Join Pushdown will not happen. 
+
+#### 2. Invoke GroupInnerJoinsByConnector optimizer based on session flag
+
+GroupInnerJoinsByConnector optimizer will be invoked based on session flag 'optimizer.inner-join-pushdown-enabled'. 
+
+If the flag is set ('optimizer-inner-join-pushdown-enabled=true'), then it's optimize method will be invoked.
+If it is not set ('optimizer-inner-join-pushdown-enabled=false' or the flag is not set) then it's optimize method will not be invoked.
+
```

Comment:
There were one more session flag that we required to restrict non-equijoin pushdown. Please add its details too

Reply:
Is that currently implemented ?

Reply:
added